File: TempoCore/Tests/TempoCoreTests/SelfReportSurveyTests.swift
// 앱 내 자기보고 설문 테스트 — 문항 계약과 채점이 웹 쪽과 어긋나지 않게 막는다.

import XCTest
@testable import TempoCore

final class SelfReportSurveyTests: XCTestCase {

    /// 필수 14 + 선택 3 = 17. 개수가 바뀌면 "2분" 약속과 분석 사양이 같이 흔들린다.
    func testQuestionCount() {
        XCTAssertEqual(SelfReportSurvey.requiredQuestionIDs.count, 14)
        XCTAssertEqual(SelfReportSurvey.optionalQuestions.count, 3)
        XCTAssertEqual(SelfReportSurvey.allQuestionIDs.count, 17)
    }

    /// P1이 P2보다 먼저여야 한다 — 뒤집으면 통념이 P1을 오염시킨다.
    func testPhaseOrderIsFixed() {
        XCTAssertEqual(SelfReportSurvey.phaseQuestions.map(\.id), ["P1", "P2"])
    }

    /// Q8은 역문항이라 제거 불가 — 이진 척도에서 유일한 무성의 판별 장치다.
    func testReverseItemExists() {
        XCTAssertTrue(SelfReportSurvey.symptomQuestions.contains { $0.id == "Q8" })
    }

    /// 증상 앵커는 P2 응답에서 온다. 고정 문구를 쓰면 위상을 가정한 채 위상 이질성을 재게 된다.
    func testAnchorUsesP2Answer() {
        let line = SelfReportSurvey.symptomAnchorLine(p2: "before")
        XCTAssertTrue(line.contains("다음 생리 오기 일주일쯤 전"))
    }

    /// "딱히 없어요"·"모르겠어요"·미응답은 폴백 문구로 — 앵커를 지어내지 않는다.
    func testAnchorFallback() {
        for value in ["none", "unknown", nil] {
            let line = SelfReportSurvey.symptomAnchorLine(p2: value)
            XCTAssertTrue(line.contains("그나마 힘들었던 때"))
        }
    }

    func testModalityRawRange() {
        let emotionalHeavy = ["Q1": "worse", "Q2": "worse", "Q3": "worse",
                              "Q4": "same", "Q5": "same", "Q6": "same", "Q9": "much"]
        XCTAssertEqual(SelfReportScoring.score(emotionalHeavy).modalityRaw, 6)

        let bodilyHeavy = ["Q1": "same", "Q2": "same", "Q3": "same",
                           "Q4": "worse", "Q5": "worse", "Q6": "worse", "Q9": "much"]
        XCTAssertEqual(SelfReportScoring.score(bodilyHeavy).modalityRaw, -6)
    }

    /// 중간 선택지(2026-09-04) — 「조금 그래요」는 문항당 1점. 0으로 접히면 답이 사라진다.
    func testSomewhatCountsAsHalf() {
        let mid = ["Q1": "somewhat", "Q2": "somewhat", "Q3": "somewhat",
                   "Q4": "same", "Q5": "same", "Q6": "same", "Q9": "much"]
        XCTAssertEqual(SelfReportScoring.score(mid).modalityRaw, 3)

        let mixed = ["Q1": "worse", "Q2": "somewhat", "Q3": "same",
                     "Q4": "somewhat", "Q5": "same", "Q6": "same", "Q9": "much"]
        XCTAssertEqual(SelfReportScoring.score(mixed).modalityRaw, 2)   // (2+1+0) − (1+0+0)
    }

    /// 루바토가 진폭 판정보다 먼저다 — 모르겠다고 답한 사람을 비바체로 밀지 않는다.
    func testRubatoTakesPrecedence() {
        XCTAssertEqual(SelfReportScoring.score(["C1": "unknown", "Q9": "total"]).type, .rubato)
        XCTAssertEqual(SelfReportScoring.score(["C1": "within1m", "Q9": "varies"]).type, .rubato)
    }

    func testAmplitudeSplit() {
        XCTAssertEqual(SelfReportScoring.score(["C1": "within1m", "Q9": "much"]).type, .vivace)
        XCTAssertEqual(SelfReportScoring.score(["C1": "within1m", "Q9": "total"]).type, .vivace)
        XCTAssertEqual(SelfReportScoring.score(["C1": "within1m", "Q9": "same"]).type, .andante)
        XCTAssertEqual(SelfReportScoring.score(["C1": "within1m", "Q9": "slight"]).type, .andante)
    }

    /// Q1~Q7 전부 "심해져요" + 역문항 Q8도 "심해져요" = 모순 → 무성의로 본다.
    func testStraightLiningDetection() {
        var all = [String: String]()
        for id in ["Q1", "Q2", "Q3", "Q4", "Q5", "Q6", "Q7", "Q8"] { all[id] = "worse" }
        XCTAssertTrue(SelfReportScoring.isStraightLining(all))

        all["Q8"] = "same"   // 역문항에서 방향을 바꿨으면 성실한 응답
        XCTAssertFalse(SelfReportScoring.isStraightLining(all))
    }

    /// 선택지 value가 중복되면 저장값이 뭉개진다.
    func testChoiceValuesAreUniquePerQuestion() {
        let all = [SelfReportSurvey.calibration]
            + SelfReportSurvey.phaseQuestions
            + SelfReportSurvey.symptomQuestions
            + SelfReportSurvey.amplitudeQuestions
            + SelfReportSurvey.optionalQuestions
        for question in all {
            let values = Set(question.choices.map(\.value))
            XCTAssertEqual(values.count, question.choices.count, "중복 value: \(question.id)")
            XCTAssertFalse(question.choices.isEmpty, "선택지 없음: \(question.id)")
        }
    }
}

File: android/tempocore/src/test/kotlin/app/temporoutine/core/SelfReportSurveyTests.kt
// 앱 내 자기보고 설문 테스트 — 문항 계약과 채점이 웹 쪽과 어긋나지 않게 막는다.
// iOS TempoCoreTests/SelfReportSurveyTests.swift 1:1 이식.

package app.temporoutine.core

import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class SelfReportSurveyTests {

    @Test fun testQuestionCount() {
        assertEquals(14, SelfReportSurvey.requiredQuestionIDs.size)
        assertEquals(3, SelfReportSurvey.optionalQuestions.size)
        assertEquals(17, SelfReportSurvey.allQuestionIDs.size)
    }

    @Test fun testPhaseOrderIsFixed() {
        assertEquals(listOf("P1", "P2"), SelfReportSurvey.phaseQuestions.map { it.id })
    }

    @Test fun testReverseItemExists() {
        assertTrue(SelfReportSurvey.symptomQuestions.any { it.id == "Q8" })
    }

    @Test fun testAnchorUsesP2Answer() {
        val line = SelfReportSurvey.symptomAnchorLine("before")
        assertTrue(line.contains("다음 생리 오기 일주일쯤 전"))
    }

    @Test fun testAnchorFallback() {
        for (value in listOf("none", "unknown", null)) {
            val line = SelfReportSurvey.symptomAnchorLine(value)
            assertTrue(line.contains("그나마 힘들었던 때"))
        }
    }

    /** P1·P2는 복수 응답(2026-09-09) — 저장은 쉼표로 이은 한 문자열이다. */
    @Test fun testPhaseQuestionsAllowMultiple() {
        assertTrue(SelfReportSurvey.phaseQuestions.all { it.allowsMultiple })
        assertFalse(SelfReportSurvey.symptomQuestions.any { it.allowsMultiple })
        assertEquals(listOf("menstrual", "before"), SelfReportSurvey.values("menstrual,before"))
        assertEquals(emptyList<String>(), SelfReportSurvey.values(null))
        assertEquals(emptyList<String>(), SelfReportSurvey.values(""))
    }

    /** 복수로 고른 위상은 앵커에 전부 나열한다 — 하나만 골라 쓰면 나머지가 화면에서 사라진다. */
    @Test fun testAnchorListsEveryPickedPhase() {
        val line = SelfReportSurvey.symptomAnchorLine("menstrual,before")
        assertTrue(line.contains("생리 중"))
        assertTrue(line.contains("다음 생리 오기 일주일쯤 전"))
        // 배타 값이 섞여 들어와도 문구에는 실제 위상만 남는다
        val mixed = SelfReportSurvey.symptomAnchorLine("menstrual,unknown")
        assertTrue(mixed.contains("생리 중"))
        assertFalse(mixed.contains("잘 모르겠어요"))
    }

    /** Q9는 3지(2026-09-09) — 선택지가 곧 유형이다. 늘리면 판정 경계가 다시 흐려진다. */
    @Test fun testAmplitudeChoicesMapOneToOne() {
        val q9 = SelfReportSurvey.amplitudeQuestions.firstOrNull { it.id == "Q9" }
        assertEquals(listOf("same", "much", "varies"), q9?.choices?.map { it.value })
    }

    @Test fun testModalityRawRange() {
        val emotionalHeavy = mapOf("Q1" to "worse", "Q2" to "worse", "Q3" to "worse",
            "Q4" to "same", "Q5" to "same", "Q6" to "same", "Q9" to "much")
        assertEquals(6, SelfReportScoring.score(emotionalHeavy).modalityRaw)

        val bodilyHeavy = mapOf("Q1" to "same", "Q2" to "same", "Q3" to "same",
            "Q4" to "worse", "Q5" to "worse", "Q6" to "worse", "Q9" to "much")
        assertEquals(-6, SelfReportScoring.score(bodilyHeavy).modalityRaw)
    }

    /** 중간 선택지(2026-09-04) — 「그때그때 달라요」(somewhat)는 문항당 1점. 0으로 접히면 답이 사라진다.
     *  2026-09-09 라벨이 바뀌어도 채점은 1점 유지(대표님 결정). */
    @Test fun testSomewhatCountsAsHalf() {
        val mid = mapOf("Q1" to "somewhat", "Q2" to "somewhat", "Q3" to "somewhat",
            "Q4" to "same", "Q5" to "same", "Q6" to "same", "Q9" to "much")
        assertEquals(3, SelfReportScoring.score(mid).modalityRaw)
        val mixed = mapOf("Q1" to "worse", "Q2" to "somewhat", "Q3" to "same",
            "Q4" to "somewhat", "Q5" to "same", "Q6" to "same", "Q9" to "much")
        assertEquals(2, SelfReportScoring.score(mixed).modalityRaw)
        // 증상 문항 선택지는 3개(강도 내림차순) — 옛 응답(worse·same)은 값이 그대로라 계속 읽힌다
        for (q in SelfReportSurvey.symptomQuestions) {
            assertEquals(listOf("worse", "somewhat", "same"), q.choices.map { it.value })
        }
    }

    @Test fun testRubatoTakesPrecedence() {
        assertEquals(RhythmType.RUBATO, SelfReportScoring.score(mapOf("C1" to "unknown", "Q9" to "total")).type)
        assertEquals(RhythmType.RUBATO, SelfReportScoring.score(mapOf("C1" to "within1m", "Q9" to "varies")).type)
    }

    /** `total`·`slight`는 화면에서 내린 옛 값이다(2026-09-09 Q9 축소) — 그 전에 답한 기록이
     *  유형 없이 떨어지지 않게 채점 집합에는 남아 있어야 한다. */
    @Test fun testAmplitudeSplit() {
        assertEquals(RhythmType.VIVACE, SelfReportScoring.score(mapOf("C1" to "within1m", "Q9" to "much")).type)
        assertEquals(RhythmType.VIVACE, SelfReportScoring.score(mapOf("C1" to "within1m", "Q9" to "total")).type)
        assertEquals(RhythmType.ANDANTE, SelfReportScoring.score(mapOf("C1" to "within1m", "Q9" to "same")).type)
        assertEquals(RhythmType.ANDANTE, SelfReportScoring.score(mapOf("C1" to "within1m", "Q9" to "slight")).type)
    }

    /** Q1~Q7 전부 "예"(worse) + 역문항 Q8도 "예" = 모순 → 무성의로 본다. */
    @Test fun testStraightLiningDetection() {
        val all = mutableMapOf<String, String>()
        for (id in listOf("Q1", "Q2", "Q3", "Q4", "Q5", "Q6", "Q7", "Q8")) all[id] = "worse"
        assertTrue(SelfReportScoring.isStraightLining(all))

        all["Q8"] = "same"
        assertFalse(SelfReportScoring.isStraightLining(all))
    }

    @Test fun testChoiceValuesAreUniquePerQuestion() {
        val all = listOf(SelfReportSurvey.calibration) +
            SelfReportSurvey.phaseQuestions +
            SelfReportSurvey.symptomQuestions +
            SelfReportSurvey.amplitudeQuestions +
            SelfReportSurvey.optionalQuestions
        for (question in all) {
            val values = question.choices.map { it.value }.toSet()
            assertEquals(question.choices.size, values.size, "중복 value: ${question.id}")
            assertFalse(question.choices.isEmpty(), "선택지 없음: ${question.id}")
        }
    }
}
